java/garbage-collector/src/main/java/sleeper/garbagecollector/GarbageCollector.java

```diff
@@ -70,12 +71,15 @@ public void run() throws StateStoreException, IOException {
             StateStore stateStore = stateStoreProvider.getStateStore(tableProperties);
 
             LOGGER.debug("Requesting iterator of files ready for garbage collection from state store");
-            Iterator<FileInfo> readyForGC = stateStore.getReadyForGCFiles();
-
+            int delayBeforeDeletion = tableProperties.getInt(GARBAGE_COLLECTOR_DELAY_BEFORE_DELETION);
+            Instant deletionTime = startTime.minus(delayBeforeDeletion, ChronoUnit.MINUTES);
+            List<String> readyForGCFilenames = stateStore.getReadyForGCFilenamesBefore(deletionTime).collect(Collectors.toList());
             int numberDeleted = 0;
-            while (readyForGC.hasNext() && numberDeleted < garbageCollectorBatchSize) {
-                FileInfo fileInfo = readyForGC.next();
-                deleteFileAndUpdateStateStore(fileInfo, stateStore, conf);
```

Comment:
Previously this deliberately avoided collecting all the files into a list. We could still do that. The easiest thing would be just to call `Stream.iterator` and use the code as it was before.

Reply:
Thanks, I've reverted to using an iterator here instead of collecting to a list